src/main/java/appeng/helpers/PatternHelper.java

```diff
@@ -85,7 +85,13 @@ public PatternHelper( final ItemStack is, final World w )
 
 		for( int x = 0; x < inTag.tagCount(); x++ )
 		{
-			final ItemStack gs = new ItemStack( inTag.getCompoundTagAt( x ) );
+			NBTTagCompound ingredient = inTag.getCompoundTagAt( x );
+			final ItemStack gs = new ItemStack( ingredient );
+
+			if( !ingredient.hasNoTags() && gs.isEmpty() )
+			{
+				throw new IllegalArgumentException( "No pattern here!" );
```

Comment:
Yes, the input side can have empty itemstacks, even all 9 can be empty as long as there is at least 1 output item. It probably will not work with assemblers, but some custom machine could use it. (But there are better options for it).

The change certainly looks valid, but processing patterns are a bit less restrictive compared to crafting ones. So there might be a reason why it was done in this way. E.g. a couple of broken mods having some strange behaviour and need it for whatever reason. Or EC2 was able to abuse it for fluid crafting etc.